### PR TITLE
Issue 1619 conftest assert rewrite

### LINKED ISSUE #1619: conftest files are not using assertion rewrite
As discovered during the sprint, we noticed that `contest.py` files are using assert reinterpretation.

### PR DESCRIPTION
**Pull request by nicoddemus**
This change now makes rewrites assertions in `conftest.py` files so they can benefit from assertion rewriting. 

Fixes #1619
cc @flub @sober7

### CODE REVIEW

File: _pytest/assertion/rewrite.py [hunk 1]
```diff
@@ -50,14 +50,14 @@ def __init__(self):
         self._register_with_pkg_resources()
 
     def set_session(self, session):
-        self.fnpats = session.config.getini("python_files")
         self.session = session
 
+    def set_config(self, config):
```
Comment (flub) [inline _pytest/assertion/rewrite.py]: Should we change this to take config in the constructor?

File: _pytest/assertion/__init__.py [hunk 2]
```diff
@@ -154,7 +164,8 @@ def _load_modules(mode):
         from _pytest.assertion import rewrite  # noqa
 
 
-def warn_about_missing_assertion(mode):
+def warn_about_missing_assertion(mode, pluginmanager):
+    print('got here')
```
Comment (The-Compiler) [inline _pytest/assertion/__init__.py]: I think this shouldn't be here?

File: _pytest/assertion/__init__.py [hunk 3]
```diff
@@ -166,10 +177,18 @@ def warn_about_missing_assertion(mode):
...
+        try:
+            out, err = capman.suspendcapture()
+            sys.stderr.write("WARNING: " + specifically +
+                             " because assert statements are not executed "
+                             "by the underlying Python interpreter "
+                             "(are you using python -O?)\n")
+        finally:
+            capman.resumecapture()
+            sys.stdout.write(out)
+            sys.stderr.write(err)
```
Comment (The-Compiler) [inline _pytest/assertion/__init__.py]: What's the point of writing the captured data out again after resuming capturing? Wouldn't it be captured and double then?

File: _pytest/assertion/rewrite.py [hunk 4]
```diff
@@ -86,24 +84,11 @@ def find_module(self, name, path=None):
...
-                        break
-                else:
-                    return None
-            finally:
-                self.session = sess
-        else:
-            state.trace("matched test file (was specified on cmdline): %r" %
-                        (fn,))
+        if not self._should_rewrite(fn_pypath, state):
+            return
```
Comment (The-Compiler) [inline _pytest/assertion/rewrite.py]: I think this should be an explicit "return None" as the other returns are explicit too

Comment (The-Compiler) [inline _pytest/assertion/rewrite.py]: I think this should be an explicit "return None" as the other returns are explicit too

File: _pytest/assertion/rewrite.py [hunk 5]
```diff
@@ -151,6 +136,32 @@ def find_module(self, name, path=None):
...
+                session = self.session  # avoid a cycle here
+                self.session = None
+                try:
+                    for pat in self.fnpats:
+                        if fn_pypath.fnmatch(pat):
+                            state.trace("matched test file %r" % (fn,))
+                            return True
+                finally:
+                    self.session = session
+                    del session
```
Comment (The-Compiler) [inline _pytest/assertion/rewrite.py]: This seems superfluous as the function returns anyways, so `session` gets out of scope

File: _pytest/assertion/__init__.py (hunk 2 above)
Comment (nicoddemus) [inline _pytest/assertion/__init__.py]: Removed, thanks!

File: _pytest/assertion/__init__.py (hunk 3 above)
Comment (nicoddemus) [inline _pytest/assertion/__init__.py]: Not really, when `suspendcapture` is called, it returns the captured output; we then print our warning, resume capture and write out again the captured data, so it doesn't get lost. We could print the captured data before or after the warning, but decided to keep the capture behavior.

File: _pytest/assertion/rewrite.py (hunk 4 above)
Comment (nicoddemus) [inline _pytest/assertion/rewrite.py]: Done, thanks

File: _pytest/assertion/rewrite.py (hunk 5 above)
Comment (nicoddemus) [inline _pytest/assertion/rewrite.py]: i think the point is to remove `session` from the traceback if an exceotion happens inside the `try/finally` statement. Not sure I know why it was here in the first place, we just kept the same code.

Comment (The-Compiler) [inline _pytest/assertion/rewrite.py]: I don't see a `del session` in the old code, but your explanation makes sense :laughing:

File: _pytest/assertion/__init__.py (hunk 3 above)
Comment (The-Compiler) [inline _pytest/assertion/__init__.py]: Are you sure it would get lost? Does `suspendcapture` delete the captured data?

Comment (nicoddemus) [inline _pytest/assertion/__init__.py]: Yes, that's why it returns it. :grin:

**Comment by nicoddemus** [comment]
Tested it on my windows laptop since AppVeyor is stuck (py27,py35,linting).

**Pull request merged**